x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -1053,63 +1058,84 @@ public void expandIds(
+                                ExceptionsHelper.serverError(
+                                    "Searching for stats for models [{}] failed",
+                                    response.getFailure(),
+                                    Strings.arrayToCommaDelimitedString(modelIds)
+                                )
+                            );
+                            return;
                         }
-                        logger.error(
-                            () -> "[" + Strings.arrayToCommaDelimitedString(modelIds) + "] search failed for models",
-                            response.getFailure()
-                        );
-                        listener.onFailure(
-                            ExceptionsHelper.serverError(
-                                "Searching for stats for models [{}] failed",
-                                response.getFailure(),
-                                Strings.arrayToCommaDelimitedString(modelIds)
-                            )
-                        );
-                        return;
-                    }
-                    try {
-                        InferenceStats inferenceStats = handleMultiNodeStatsResponse(response.getResponse(), modelIds[modelIndex++]);
-                        if (inferenceStats != null) {
-                            allStats.add(inferenceStats);
+                        try {
+                            InferenceStats inferenceStats = handleMultiNodeStatsResponse(response.getResponse(), modelIds[modelIndex++]);
+                            if (inferenceStats != null) {
+                                allStats.add(inferenceStats);
+                            }
+                        } catch (Exception e) {
+                            listener.onFailure(e);
+                            return;
                         }
-                    } catch (Exception e) {
-                        listener.onFailure(e);
+                    }
+                    listener.onResponse(allStats);
+                }, e -> {
+                    Throwable unwrapped = ExceptionsHelper.unwrapCause(e);
+                    if (unwrapped instanceof ResourceNotFoundException) {
+                        listener.onResponse(Collections.emptyList());
                         return;
                     }
-                }
-                listener.onResponse(allStats);
-            }, e -> {
-                Throwable unwrapped = ExceptionsHelper.unwrapCause(e);
-                if (unwrapped instanceof ResourceNotFoundException) {
-                    listener.onResponse(Collections.emptyList());
-                    return;
-                }
-                listener.onFailure((Exception) unwrapped);
-            }),
-            client::multiSearch
-        );
+                    listener.onFailure((Exception) unwrapped);
+                }),
+                client::multiSearch
+            );
+
+        }).addListener(listener);
```

Comment:
I think this would also need to be moved back, unfortunately:

```suggestion
        })
            .addListener(
                listener,
                client.threadPool().executor(MachineLearning.UTILITY_THREAD_POOL_NAME),
                client.threadPool().getThreadContext()
            );
```

Reply:
ah, interesting!